### ISSUE TITLE
Add in stochastic movement to enemy AI

### ISSUE THREAD

**Issue by JosriqueJotaro** [Micky774/unity_project#10]
I believe that we should consider adding in some sort of stochastic movement to the enemy AI. The current implementation of this stochastic movement is in the ai_toy branch.

In the current implementation, at each update step, Enemy.cs calculates a random vector, called `jitter_velocity_change` using the `GetPointOnCircleMethod`:
```
Vector2 GetPointOnCircle(float radius = 1) {
    float random_angle = Random.Range(0f, 2 * Mathf.PI - Mathf.Epsilon);
    Vector2 point_on_circle = new Vector2(Mathf.Cos(random_angle), Mathf.Sin(random_angle)) * radius;
    return point_on_circle;
}   
```

This method is employed in the `Update` function as follows:
```
Vector2 jitter_velocity_change = GetPointOnCircle(jitter_rate);
```
`jitter_velocity_change` is added to the previous `jitter_velocity` (which is initialized as 0) and clamped to a max jitter_speed:
```
jitter_velocity = Vector2.ClampMagnitude(jitter_velocity + jitter_velocity_change, max_jitter_speed);
```
A separate, independent variable, `productive_velocity', calculates the velocity change which directs the enemy AI toward the player. At the end of `Update`, the productive and jitter velocities are added together.

The issue with this current implementation is that the jitter is... jittery. It does not look particularly aesthetically pleasing to me. Further, it should be known that jitter might be alright for some varieties of enemies, but for the floaty bois, the jitter doesn't make a ton of sense. Perhaps, for the floaty bois, it would make more sense for them to have some sort of stochastic curvature - this would make it such that the path toward the player is C2 and makes more sense for a flying character.

**Comment by Micky774** [comment]
I do not think that we need noisy paths for the simple tracking behaviors, however I do believe some form of stochastic variation in paths/patterns would be a welcome addition. In fact the simplest could be to randomize attack _timings_ (e.g. a charge could last between 1-3 seconds).

A few general notes on noise and randomness:
1. We can talk about "discrete" and "continuous" noise, such as rolling a die or sampling the motion of a random particle in a box.
2. We can talk about "categorical" and "smooth" features, such as an attack pattern chosen from finitely-many unique options, or a trajectory in space/time (i.e. accounting for velocity).
3. We can talk about independent noise (e.g. brownian motion) and autocorrellated noise (e.g. metropolis-hastings sampling).

A _very rough rule of thumb_ is as follows: smooth features benefit from autocorrellated noise generated by continious distributions. Categorical features benefit from independent noise from discrete distributions. From that lens, the jitter you implemented initially is independently sampled noise from a continous distribution. The excess jitter and aesthetic gap is due to this lack of autocorrellation.

An example of a similar, yet auto-correllated noise would be to, instead of uniformally sampling the angle across the sphere, start with a random angle and sample _deviations_ within the range $\[\frac{-\pi}{4}, \frac{\pi}{4}\]$ at fixed time-steps, then linearly interpolate between them.

For example:
1. Set deviation time step $T = 10$
2. Sample starting angle $\theta\sim U(0, 2\pi)$
3. Sample target deviation $\Delta\theta\sim U(\frac{-\pi}{4}, \frac{\pi}{4})$
4. For timestep $t$, set actual angle $\hat\theta(t) = \theta + \frac{t}{T}\Delta\theta$

Again, I don't believe it would be benificial for this specific behavior (direct linear tracking) however it is probably going to be beneficial, if not outright _necessary_ for more complex behaviors, and it doesn't hurt to start thinking about it. We want to include as much randomness as is healthy since it is a free way to add complexity and replayability to the game without sacrificing consistency and intent.

**Comment by AnonymousFool** [comment]
> I would say that really it is a writing team discussion first and foremost, then art team, but yeah. I think I'll petition for maybe bi-weekly "all hands" meetings where we can sync up to bring up ideas like this.

Obviously I think it needs the approval of the writing department, but I don't think the writing department would have the most valuable input on the idea at this stage. Their job is to craft a consistent, coherent world that evokes the right feeling, sure, but I don't expect that knowledge of spooky aesthetics and animation styles are really a core part of that job, if anything that's just something I expect them to give observational feedback on, accept or reject, and incorporate if accepted. I'd expect the art department to be the one that would actually have the knowledge and experience to flesh out a currently raw animation idea and provide valuable stylistic additions.

**Comment by Micky774** [comment]
@JosriqueJotaro
